src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -1460,30 +1519,27 @@ def process_declarations(self, parent, nodes, arg_list):
                 # Restore the fparser2 parse tree
                 decl.children[2].items = tuple(orig_children)
 
-        # Check for symbols named in an access statement but not explicitly
-        # declared. These must then refer to symbols that have been brought
-        # into scope by an unqualified use statement. As we have no idea
-        # whether they represent data or a routine we use the Symbol base
-        # class.
-        for name in (list(explicit_public_symbols) +
-                     list(explicit_private_symbols)):
-            if name not in parent.symbol_table:
-                if name in explicit_public_symbols:
-                    vis = Symbol.Visibility.PUBLIC
-                else:
-                    vis = Symbol.Visibility.PRIVATE
-                # TODO 736 Ideally we would use parent.find_or_create_symbol()
-                # here since that checks that there is a possible source for
-                # this previously-unseen symbol. However, we cannot yet do this
-                # because we don't capture symbols for routine names so
-                # that, e.g.:
-                #   module my_mod
-                #     public my_routine
-                #   contains
-                #     subroutine my_routine()
-                # would cause us to raise an exception.
-                parent.symbol_table.add(Symbol(name, visibility=vis))
-
+        if visibility_map is not None:
+            # Check for symbols named in an access statement but not explicitly
+            # declared. These must then refer to symbols that have been brought
+            # into scope by an unqualified use statement.
+            for name, vis in visibility_map.items():
+                if name not in parent.symbol_table:
+                    # TODO #876 find_or_create_symbol() always creates a
+                    # DataSymbol if no matching symbol already exists and there
+                    # is at least one possible Container from which it may be
+                    # imported. This will be wrong if the symbol is actually
+                    # the name of a routine.
+                    try:
+                        sym = parent.find_or_create_symbol(name,
+                                                           visibility=vis)
+                    except SymbolError:
+                        # Improve the error message with context-specific info
+                        raise SymbolError(
```

Comment:
The new pylint suggests writing this expression like:
```
except SymbolError as e:
   raise SymbolError("...") from e
```
to keep the full error stack trace.

However this syntax is only available on python3, but here is a `six.raise_from()` that may do the same.

I will approve this either way, but I thought to mention it in case you want to investigate it.

Reply:
Ooh, I didn't know about that. I've changed to `six.raise_from()` that does what pylint suggests for python 3 and does what I originally had for python 2. Everyone wins :-)